File: app/src/main/java/com/jeketos/associatewith/guesser/GuesserActivity.java
package com.jeketos.associatewith.guesser;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.MotionEvent;
import android.widget.EditText;
import android.widget.ImageView;
import com.jeketos.associatewith.Point;
import com.jeketos.associatewith.R;
import com.jeketos.associatewith.di.Injector;
import com.jeketos.associatewith.guesser.chat.ChatAdapter;
import com.jeketos.associatewith.guesser.chat.IChatItem;
import com.jeketos.associatewith.util.DialogUtils;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by eugene.kotsogub on 10/28/16.
 *
 */

public class GuesserActivity extends AppCompatActivity implements GuesserMVP.GuesserView {

    GuesserMVP.GuesserPresenter presenter;
    private Canvas canvas;
    private Paint paint;
    @BindView(R.id.image_view)
    ImageView imageView;
    @BindView(R.id.recycler_view)
    RecyclerView chatRecyclerView;
    @BindView(R.id.edit_text)
    EditText editText;
    private ChatAdapter chatAdapter;

    @OnClick(R.id.send) void OnSendClick(){
        String message = editText.getText().toString();
        if(!TextUtils.isEmpty(message)){
            presenter.sendMessage(message);
            editText.setText(null);
        }
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_guesser);
        ButterKnife.bind(this);
        presenter = Injector.provideGuesserPresenter(this);
    }

    @Override
    public void init() {
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
        Bitmap bitmap = Bitmap.createBitmap(displayMetrics.widthPixels,displayMetrics.heightPixels,
                Bitmap.Config.ARGB_8888);
        canvas = new Canvas(bitmap);
        paint = new Paint();
        paint.setColor(Color.BLACK);
        paint.setAntiAlias(true);
        paint.setDither(true);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStrokeWidth(12);
        imageView.setImageBitmap(bitmap);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        chatRecyclerView.setLayoutManager(layoutManager);
        chatAdapter = new ChatAdapter();
        chatRecyclerView.setAdapter(chatAdapter);
    }

    @Override
    public void draw(float previousX, float previousY, Point point) {
        switch (point.getMotionEvent()){
            case MotionEvent.ACTION_DOWN:
                drawPoint(point);
                break;
            default:
                drawLine(previousX, previousY, point);
                break;
        }
    }

    @Override
    public void clearBoard() {
        init();
    }

    @Override
    public void clearChat() {
        chatAdapter.updateItems(null);
    }

    @Override
    public void addChatItem(IChatItem item) {
        chatAdapter.updateItems(item);
    }

    @Override
    public void showWinnerDialog(String name, String word) {
        AlertDialog.Builder builder = DialogUtils.createWinnerDialog(this, name, word);
        builder.create();
        builder.show();
    }


    public void drawPoint(Point point) {
        canvas.drawPoint(point.getX(), point.getY(), paint);
        imageView.invalidate();
    }


    public void drawLine(float previousX, float previousY, Point point) {
        canvas.drawLine(previousX, previousY, point.getX(), point.getY(), paint);
        imageView.invalidate();
    }
}

File: app/src/main/java/com/jeketos/associatewith/guesser/GuesserActivity.kt
package com.jeketos.associatewith.guesser

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.text.TextUtils
import android.view.MotionEvent
import android.view.inputmethod.InputMethodManager
import com.jeketos.associatewith.Point
import com.jeketos.associatewith.R
import com.jeketos.associatewith.base.BaseActivity
import com.jeketos.associatewith.base.BasePresenter
import com.jeketos.associatewith.drawer.DrawerActivity
import com.jeketos.associatewith.guesser.chat.ChatAdapter
import com.jeketos.associatewith.guesser.chat.IChatItem
import com.jeketos.associatewith.util.DialogUtils
import kotlinx.android.synthetic.main.activity_guesser.*
import javax.inject.Inject

/**
 * Created by eugene.kotsogub on 10/28/16.
 *
 */

class GuesserActivity() : BaseActivity<GuesserMVP.GuesserView>(), GuesserMVP.GuesserView {

    @Inject lateinit var presenter : GuesserMVP.GuesserPresenter
    lateinit var  chatAdapter : ChatAdapter

    fun onSendClick(){
         val message = editText.text.toString()
        if(!TextUtils.isEmpty(message)){
            presenter.sendMessage(message)
            editText.text = null
        }
    }

    override fun getPresenter(): BasePresenter<GuesserMVP.GuesserView> {
        return presenter
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        component.inject(this)
        setContentView(R.layout.activity_guesser)
        init()
    }

    override fun onStart() {
        super.onStart()
        presenter.init()
    }

    override fun init() {
        send.setOnClickListener {onSendClick()}
        chatRecyclerView.layoutManager = LinearLayoutManager(this)
        chatAdapter = ChatAdapter()
        chatRecyclerView.adapter = chatAdapter
    }

    override fun draw(point: Point) {
        imageView.setDrawColor(point.color)
        imageView.setStrokeWidth(point.strokeWidth)
        val motionEvent = MotionEvent.obtain(System.currentTimeMillis(), System.currentTimeMillis(), point.motionEvent, point.x, point.y, 0)
        imageView.onTouchEvent(motionEvent)
    }

    override fun clearBoard() {
        init()
        imageView.clear()
    }

    override fun setDrawColor(color: Int) {
        imageView.setDrawColor(color)
    }

    override fun clearChat() {
        chatAdapter.updateItems(null)
    }

    override fun addChatItem(item: IChatItem) {
        chatAdapter.updateItems(item)
    }

    override fun showWinnerDialog(name: String, word: String, isWinner: Boolean) {
        val builder = DialogUtils.createWinnerDialog(this, name, word)
        builder.setPositiveButton(android.R.string.ok, {d,i ->
            hideKeyboard()
            finish()
            if(isWinner) {
                startActivity(Intent(this, DrawerActivity::class.java))
            } else{
                startActivity(Intent(this, GuesserActivity::class.java))
            }
        })
        builder.create().show()
    }

    fun hideKeyboard(){
        if (getCurrentFocus() != null) {
            (getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).hideSoftInputFromWindow(getCurrentFocus()!!.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }
}
